Support Component Connections in USD

danieldresser-ie opened the pull request:
Hopefully this matches the API you were picturing based on our conversation.

The test coverage is pretty decent, and this does seem to give an exact match on my one complex production test, so this is hopefully just about good to go.

The only things that came up in production testing were unrelated, but I'm still leaving a note here to talk about them tomorrow while I'm thinking of it: the IE render preview UI requires adapting to work with the new Inspector columns on the Catalogue ( or maybe the Catalogue itself is having issues? ), and it seems like it may be possible to get unexpectedly inconsistent results when comparing precise results from the Catalogue using Catalogue Selects ( ie. if it matters whether it's been converted to half-float ).

johnhaddon commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -233,70 +228,245 @@ void ShaderNetworkAlgo::convertOSLComponentConnections( ShaderNetwork *network,
+			}
+			else
+			{
+				if( inputConnections.begin() == inputConnections.end() )
+				{
+					throw IECore::Exception( boost::str(
+						boost::format(
+							"removeComponentConnectionAdapters : mx_swizzle_color_float \"%1%\" must have an input"
+						) % s.first.string()
+					) );
 				}
```
This assumes that if there's at least one connection, the first one must be into `swizzle.in`. We could check more accurately by calling `ShaderNetwork::input()`, and then we wouldn't need `inputConnections` at all for this branch.

johnhaddon commented on src/IECoreScene/bindings/ShaderNetworkAlgoBinding.cpp:
```diff
@@ -61,6 +61,11 @@ CompoundDataPtr expandSplineParametersWrapper( CompoundDataPtr parameters )
 	return boost::const_pointer_cast< CompoundData >( ShaderNetworkAlgo::expandSplineParameters( parameters ) );
 }
 
+IECore::InternedString componentConnectionAdapterLabelWrapper()
```
Worth converting this to a regular Python string on return I think - InternedStrings are in a bit of an uncanny valley when it comes to Python.

johnhaddon commented on test/IECoreScene/ShaderNetworkAlgoTest.py:
```diff
@@ -263,6 +263,100 @@ def testArnoldComponentConnectionsNotConverted( self ) :
+		self.assertEqual( convertBack, network )
+
+		# Check that the shaders we don't modify are reused when the network is copied ( the addresses are the same )
+		self.assertEqual( str( network.getShader( "source1" ) ), str( converted.getShader( "source1" ) ) )
+		self.assertEqual( str( network.getShader( "dest" ) ), str( converted.getShader( "dest" ) ) )
```
I don't understand this. For RefCounted objects, `a == b` or `a.isSame( b )` would check identity accurately. And I think it would fail here, because `ShaderNetwork.getShader()` returns a copy to Python (so you can't accidentally modify the thing that was const in C++). Which makes me thing that the string comparison is probably bogus?

In several places in Gaffer (e.g. `TypedObjectPlug.getValue()`), we use an optional `_copy` argument to control the behaviour of functions like `getShader`, so perhaps we should add that and then do this?

```
self.assertEqual( network.getShader( "source1", _copy = False ), converted.getShader( "source1", _copy = False ) )
```

danieldresser-ie replied:
Oh, fair.  I didn't really think too much about this one, just saw that the string representation contained an address, and it matched, which seemed right, and figured it was good.   I guess you're saying that the Python wrapper is probably allocating a new object at this address, and because it immediately is discarded, the next copy object ends up at the same address?

If this was independent code, it seems like it would be healthier to just add a "getAddress" to be used for debugging, rather than "_copy = False", which seems more confusing and harmful, if only needed for debugging, but I guess "_copy = False" is the convention.

At the end, you mean I should use `self.assertTrue( network.getShader( ... ).isSame( network.getShader( ... ) )`, right?  The _copy wouldn't do anything if we were just doing value comparison?

johnhaddon replied:
> At the end, you mean I should use self.assertTrue( network.getShader( ... ).isSame( network.getShader( ... ) ), right? The _copy wouldn't do anything if we were just doing value comparison?

Oh, yep, that looks good. I was thinking `assertEqual()` would do the job because the RefCounted bindings compare by pointer, but the Object bindings override that to compare values. So yep, `assertTrue( a.isSame( b ) )` is it.

johnhaddon reviewed (changes requested):
Thanks Daniel, LGTM apart from the small comments to prove that I have actually read it all :) A couple of them are probably worth addressing though?

danieldresser-ie commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -233,70 +228,245 @@ void ShaderNetworkAlgo::convertOSLComponentConnections( ShaderNetwork *network,
+	ParameterMap outputConversions;
+	for( const auto &s : network->shaders() )
+	{
+		if( s.second->getType() != "osl:shader" )
+		{
+			continue;
+		}
```
I've currently left it, but grouped it below with the conditional that checks if the adapter is valid and prints a warning otherwise.

danieldresser-ie commented on src/IECoreScene/ShaderNetworkAlgo.cpp:
```diff
@@ -233,70 +228,245 @@ void ShaderNetworkAlgo::convertOSLComponentConnections( ShaderNetwork *network,
-					network->addConnection( { { packHandle, g_outParameterName }, { shader.first, parameterName } } );
+		if( !isSwizzle && !isPack )
+		{
+			continue;
```
I debated whether this should be a warning or exception, but now that the label is hardcoded cortex_autoAdapter, it's pretty unreasonable that would show up anywhere unexpected, and I've gone with an exception.

danieldresser-ie commented:
Added a commit with the new `_copy = False` parm for ShaderNetworkBinding, and a fixes commit that addresses your comments for you to glance at before I squash.

johnhaddon commented:
Thanks Daniel - LGTM. I've pushed a squashed version so I can merge and get a release out today.
